### ISSUE TITLE
Message boxes hang on Windows

### ISSUE THREAD

**Issue by branpk** [rust-windowing/winit#1698]
Using a different library to open a message box after creating a winit event loop results in a hang.

```rust
fn main() {
    let _event_loop = winit::event_loop::EventLoop::new();

    tinyfiledialogs::message_box_ok("", "", tinyfiledialogs::MessageBoxIcon::Info);
    panic!("Doesn't reach here");
}
```

I've tested this with both tinyfiledialogs and nfd on Windows 10. Other dialogues like open/save file work fine (except when an existing file is chosen, since it brings up a confirmation message box).

This is on winit master d103dc2. It seems like 0.22.2 is okay with the message box, but panics afterward due to #1611.

**Comment by maroider** [comment]
Does your program hang indefinitely if you do the following instead?
```Rust
fn main() {
    let event_loop = winit::event_loop::EventLoop::new();
    event_loop.run(|_, _, control_flow| {
        tinyfiledialogs::message_box_ok("", "", tinyfiledialogs::MessageBoxIcon::Info);
        *control_flow = ControlFlow::Exit;
    })
}
```

**Comment by branpk** [comment]
Yep. It also hangs regardless of whether or not a winit window is open and handling events at the same time.

**Comment by chrisduerr** [comment]
I'd assume the dialog itself actually works?  So you can interact with it and click it?

**Comment by branpk** [comment]
The message box actually doesn't display at all on master (sorry, should have mentioned that). In addition, it seems to even lock up the program that it was run from (VSCode, cmd, etc).

On 0.22.2 the message box displays and you can interact with it, but winit panics after it closes.

**Comment by maroider** [comment]
I did some bisecting, and dd866a7 (#1615) introduces the hang and makes the message box not show itself, which is consistent with what you've reported, @branpk.
It seems like things were working correctly up until the Windows event loop refactor in #1469 (b4c6cdf) got merged, at which point message boxes began making `winit` panic with `thread 'main' panicked at 'either event handler is re-entrant (likely), or no event handler is registered (very unlikely)', D:\winit\src\platform_impl\windows\event_loop\runner.rs:236:18`.

The lock-up of parent applications that happens when you experience the hang is likely due to how `tinyfiledialogs` uses `MessageBoxW(GetForegroundWindw(), ...)` instead of `MessageBoxW(ptr::null_mut(), ...)`. `native-dialog` uses the second form, and does not trigger the lock-up.

**Comment by maroider** [comment]
I tried attaching a debugger after the bug was triggered and it seems like it always gets stuck inside `GetMessageW` [here](https://github.com/rust-windowing/winit/blob/d612a1b5a10cc63413b0d7026eedcdab78d2d181/src/platform_impl/windows/event_loop.rs#L323) when using `native-dialog` and `tinyfiledialogs`. The dialog library always gets stuck inside `MessageBoxW`. 

A temporary fix would be to spawn your message boxes on a background thread.

**Comment by qthree** [comment]
I had similar problem with `msgbox` crate and solved it with a workaround from `StackOverflow` site: https://github.com/bekker/msgbox-rs/pull/16

**Comment by maroider** [comment]
Pressing <kbd>Alt</kbd> seems to make the message box appear. This requires that you either don't change the focus after creating the message box or have a Winit window in focus.

**Comment by maroider** [comment]
Removing the `WS_VISIBLE` style from the "[event target window](https://github.com/rust-windowing/winit/blob/66c117e599209ca1fb95219669ed161c1d3cc3ed/src/platform_impl/windows/event_loop.rs#L589)" seems to make message boxes behave correctly, but that probably breaks something elsewhere.

**Comment by maroider** [comment]
This issue didn't get linked from #1769, which seems to fix the issue for me.
That PR just got merged, so could you try using the latest changes on the master branch, @qthree & @branpk?

**Comment by VZout** [comment]
Ha I got good scare when I saw this in my PR:
![chrome_IDtYEKCb3l](https://user-images.githubusercontent.com/5305297/100396993-cf14f300-3047-11eb-9aaf-22116a8170df.png)

I thought I broke message boxes for some one else somehow😅

**Comment by branpk** [comment]
The bug seems resolved on master and 0.24.0. Thanks all!
